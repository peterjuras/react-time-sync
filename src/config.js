export function hasConfigChanged(oldConfig, newConfig) {
  if (!oldConfig) {
    return true;
  }

  const oldKeys = Object.keys(oldConfig);
  if (oldKeys.length !== Object.keys(newConfig).length) {
    return true;
  }

  return oldKeys.some(key => newConfig[key] !== oldConfig[key]);
}
