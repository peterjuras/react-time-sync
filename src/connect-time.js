import React, { Component } from 'react';

import PropTypes from 'prop-types';

const DEFAULT_COMPONENT_CONFIG = {
  timeProp: 'currentTime'
};

function validateComponentConfig(componentConfig) {
  if (
    typeof componentConfig.timeProp !== 'undefined' &&
    (
      typeof componentConfig.timeProp !== 'string' ||
      componentConfig.timeProp === ''
    )
  ) {
    throw new Error('timeProp must be a non-empty string value');
  }
}

export function connectTime(timerConfig, componentConfig = {}) {
  validateComponentConfig(componentConfig);

  const usedTimerConfig = timerConfig || {};
  const usedComponentConfig = {
    ...DEFAULT_COMPONENT_CONFIG,
    ...componentConfig
  };

  return (WrappedComponent) => {
    class TimeComponent extends Component {
      static contextTypes = {
        timeSync: PropTypes.object
      };

      constructor(props, context) {
        super(props);

        if (!context.timeSync) {
          throw new Error('Warning! TimeSync context cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?');
        }

        this.state = {
          [usedComponentConfig.timeProp]: context.timeSync.getCurrentTime(usedTimerConfig)
        };
      }

      componentDidMount() {
        this.removeTimer = this.context.timeSync.addTimer(this.onTick, usedTimerConfig);
      }

      componentWillUnmount() {
        if (this.removeTimer) {
          this.removeTimer();
        }
      }

      onTick = (currentTime) => {
        this.setState({
          [usedComponentConfig.timeProp]: currentTime
        });
      };

      render() {
        return <WrappedComponent {...this.props} {...this.state} />;
      }
    }

    return TimeComponent;
  };
}
