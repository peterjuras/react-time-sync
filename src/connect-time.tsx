import React from "react";
import { ReactComponentLike } from "prop-types";
import { useTime, TimerConfig } from "./use-time";

interface ComponentConfig {
  timeProp: string;
}

const DEFAULT_COMPONENT_CONFIG: ComponentConfig = {
  timeProp: "currentTime",
};

function validateComponentConfig(
  componentConfig: Partial<ComponentConfig>,
): void {
  if (
    typeof componentConfig.timeProp !== "undefined" &&
    (typeof componentConfig.timeProp !== "string" ||
      componentConfig.timeProp === "")
  ) {
    throw new Error("timeProp must be a non-empty string value");
  }
}

export function connectTime(
  timerConfig?: TimerConfig | null,
  componentConfig: Partial<ComponentConfig> = {},
): (
  WrappedComponent: ReactComponentLike,
) => (props: Record<string, unknown>) => JSX.Element {
  validateComponentConfig(componentConfig);

  const usedTimerConfig = timerConfig || {};
  const usedComponentConfig = {
    ...DEFAULT_COMPONENT_CONFIG,
    ...componentConfig,
  };

  return (
    WrappedComponent: ReactComponentLike,
  ): ((props: Record<string, unknown>) => JSX.Element) => {
    return function TimeComponent(props: Record<string, unknown>): JSX.Element {
      const timeProps = {
        ...props,
        [usedComponentConfig.timeProp]: useTime(usedTimerConfig),
      };
      return <WrappedComponent {...timeProps} />;
    };
  };
}
