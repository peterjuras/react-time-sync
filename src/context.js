import React from "react";

export const TIMESYNC_PROP = "$$_TIME_SYNC_HIDDEN_$$";

export default React.createContext();
