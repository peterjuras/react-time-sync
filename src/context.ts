import React from "react";
import TimeSync from "time-sync";

export const TIMESYNC_PROP = "$$_TIME_SYNC_HIDDEN_$$";

export interface TimeSyncContext {
  getCurrentTime: typeof TimeSync.getCurrentTime;
  getTimeLeft: typeof TimeSync.getTimeLeft;
  addTimer: typeof TimeSync.prototype.addTimer;
  createCountdown: typeof TimeSync.prototype.createCountdown;
}

const timeSync = new TimeSync();

export default React.createContext<TimeSyncContext>({
  getCurrentTime: TimeSync.getCurrentTime,
  getTimeLeft: TimeSync.getTimeLeft,
  addTimer: timeSync.addTimer,
  createCountdown: timeSync.createCountdown,
});
