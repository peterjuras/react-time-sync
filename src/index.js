import TimeSync from 'time-sync';

export { connectTime } from './connect-time';
export { default as TimeProvider } from './time-provider';
export const {
  SECONDS,
  MINUTES,
  HOURS,
  DAYS
} = TimeSync;
