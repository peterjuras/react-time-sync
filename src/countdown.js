import React, { Component } from "react";
import PropTypes from "prop-types";
import TimeContext, { TIMESYNC_PROP } from "./context";
import { hasConfigChanged } from "./config";

function getCountdownConfig(props) {
  const newCountdownConfig = {};

  if (props.interval) {
    newCountdownConfig.interval = props.interval;
  }

  if (props.until) {
    newCountdownConfig.until = props.until;
  } else {
    newCountdownConfig.until = 0;
  }

  return newCountdownConfig;
}

class Countdown extends Component {
  static propTypes = {
    children: PropTypes.func.isRequired,
    until: PropTypes.number,
    interval: PropTypes.string
  };

  static defaultProps = {
    until: null,
    interval: null
  };

  constructor(props) {
    super(props);

    if (!props[TIMESYNC_PROP]) {
      throw new Error(
        "Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"
      );
    }
  }

  state = {};

  componentDidMount() {
    this.resetCountdown();
  }

  componentDidUpdate(prevProps, prevState) {
    const { countdownConfig } = this.state;

    if (prevState.countdownConfig !== countdownConfig) {
      this.resetCountdown();
    }
  }

  componentWillUnmount() {
    if (this.stopCountdown) {
      this.stopCountdown();
    }
  }

  onCountdownTick = timeLeft => {
    this.setState({ timeLeft });
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    const countdownConfig = getCountdownConfig(nextProps);
    const hasChanged = hasConfigChanged(
      prevState.countdownConfig,
      countdownConfig
    );

    if (hasChanged) {
      return {
        countdownConfig,
        timeLeft: nextProps[TIMESYNC_PROP].getTimeLeft(countdownConfig)
      };
    }

    return null;
  }

  resetCountdown() {
    const { countdownConfig, timeLeft } = this.state;

    if (this.stopCountdown) {
      this.stopCountdown();
      this.stopCountdown = null;
    }

    if (timeLeft === 0) {
      return;
    }

    const { [TIMESYNC_PROP]: timeSync } = this.props;

    this.stopCountdown = timeSync.createCountdown(
      this.onCountdownTick,
      countdownConfig
    );
  }

  render() {
    const { children } = this.props;
    const { timeLeft } = this.state;

    return children({
      timeLeft
    });
  }
}

export default function CountdownWrapper(props) {
  return (
    <TimeContext.Consumer>
      {timeSync => <Countdown {...{ ...props, [TIMESYNC_PROP]: timeSync }} />}
    </TimeContext.Consumer>
  );
}
