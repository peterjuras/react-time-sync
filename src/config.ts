import { TimerConfig } from "./timed";
import { SafeCountdownConfig } from "./countdown";

export function hasConfigChanged(
  oldConfig: TimerConfig | SafeCountdownConfig,
  newConfig: TimerConfig | SafeCountdownConfig
): boolean {
  const oldKeys = Object.keys(oldConfig);
  if (oldKeys.length !== Object.keys(newConfig).length) {
    return true;
  }

  return oldKeys.some(key => newConfig[key] !== oldConfig[key]);
}
