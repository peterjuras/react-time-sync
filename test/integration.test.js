import { HOURS, MINUTES } from '../src/index';

import PropTypes from 'prop-types';
import React from 'react';
import TimeProvider from '../src/time-provider';
import { connectTime } from '../src/connect-time';
import lolex from 'lolex';
import { mount } from 'enzyme';

describe('#integration', () => {
  let clock;

  beforeEach(() => {
    clock = lolex.install({ now: 1 });
  });

  afterEach(() => {
    clock.uninstall();
  });

  it('should render a mix of wrapped components', () => {
    const TimeRenderer = ({ currentTime }) => <div>{currentTime}</div>;
    TimeRenderer.propTypes = { currentTime: PropTypes.number.isRequired };

    const SecondChild = connectTime()(TimeRenderer);
    const MinuteChild = connectTime({ interval: MINUTES })(TimeRenderer);
    const HourChild = connectTime({ interval: HOURS })(TimeRenderer);

    const ref = mount((
      <TimeProvider>
        <SecondChild />
        <MinuteChild />
        <HourChild />
      </TimeProvider>
    ));

    expect(ref).toMatchSnapshot();

    clock.tick(1000 * 60 * 60);
    clock.tick(1000 * 60 * 5);
    clock.tick(1000 * 20);

    ref.update();
    expect(ref).toMatchSnapshot();
    ref.unmount();
  });
});
