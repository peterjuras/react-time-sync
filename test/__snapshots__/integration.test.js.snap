// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#integration should render a mix of wrapped components 1`] = `
<TimeProvider>
  <TimeComponent>
    <TimeRenderer
      currentTime={0}
    >
      <div>
        0
      </div>
    </TimeRenderer>
  </TimeComponent>
  <TimeComponent>
    <TimeRenderer
      currentTime={0}
    >
      <div>
        0
      </div>
    </TimeRenderer>
  </TimeComponent>
  <TimeComponent>
    <TimeRenderer
      currentTime={0}
    >
      <div>
        0
      </div>
    </TimeRenderer>
  </TimeComponent>
</TimeProvider>
`;

exports[`#integration should render a mix of wrapped components 2`] = `
<TimeProvider>
  <TimeComponent>
    <TimeRenderer
      currentTime={3920}
    >
      <div>
        3920
      </div>
    </TimeRenderer>
  </TimeComponent>
  <TimeComponent>
    <TimeRenderer
      currentTime={3900}
    >
      <div>
        3900
      </div>
    </TimeRenderer>
  </TimeComponent>
  <TimeComponent>
    <TimeRenderer
      currentTime={3600}
    >
      <div>
        3600
      </div>
    </TimeRenderer>
  </TimeComponent>
</TimeProvider>
`;
